pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,446 @@
+	// synced at least once. Added as a member to the struct to allow injection
+	// for testing.
+	servicesSynced cache.InformerSynced
+
+	// reconciler is an util used to reconcile EndpointSlice changes.
+	reconciler *reconciler
+
+	// Endpoints that need to be updated. A channel is inappropriate here,
+	// because it allows Endpoints with lots of addresses to be serviced much
+	// more often than Endpoints with few addresses; it also would cause an
+	// Endpoints resource that's inserted multiple times to be processed more
+	// than necessary.
+	queue workqueue.RateLimitingInterface
+
+	// maxEndpointsPerSubset references the maximum number of endpoints that
+	// should be added to an EndpointSlice for an EndpointSubset.
+	maxEndpointsPerSubset int32
+
+	// workerLoopPeriod is the time between worker runs. The workers process the
+	// queue of changes to Endpoints resources.
+	workerLoopPeriod time.Duration
+
+	// endpointUpdatesBatchPeriod is an artificial delay added to all Endpoints
+	// syncs triggered by EndpointSlice changes. This can be used to reduce
+	// overall number of all EndpointSlice updates.
+	endpointUpdatesBatchPeriod time.Duration
+}
+
+// Run will not return until stopCh is closed.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting EndpointSliceMirroring controller")
+	defer klog.Infof("Shutting down EndpointSliceMirroring controller")
+
+	if !cache.WaitForNamedCacheSync("endpoint_slice_mirroring", stopCh, c.endpointsSynced, c.endpointSlicesSynced, c.servicesSynced) {
+		return
+	}
+
+	klog.V(2).Infof("Starting %d worker threads", workers)
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same service
+// at the same time
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncEndpoints(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error mirroring EndpointSlices for %q Endpoints, retrying. Error: %v", key, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Retry budget exceeded, dropping %q Endpoints out of the queue: %v", key, err)
+	c.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (c *Controller) syncEndpoints(key string) error {
+	startTime := time.Now()
+	defer func() {
+		syncDuration := float64(time.Since(startTime).Milliseconds()) / float64(1000)
```

Comment:
do numeric constants need to be typed w/ float?

Reply:
Good catch, fixed.